packages/backend/src/api/controllers/ForcedTradeOfferController.ts

```diff
@@ -0,0 +1,289 @@
+
+export class ForcedTradeOfferController {
+  constructor(
+    private offerRepository: ForcedTradeOfferRepository,
+    private stateUpdateRepository: StateUpdateRepository,
+    private userRegistrationEventRepository: UserRegistrationEventRepository
+  ) {}
+
+  async postOffer(
+    offer: Omit<ForcedTradeInitialOfferRecord, 'createdAt' | 'id'>
+  ): Promise<ControllerResult> {
+    const positionA = await this.stateUpdateRepository.getPositionById(
+      offer.positionIdA
+    )
+
+    if (!positionA) {
+      return { type: 'not found', content: 'Position does not exist.' }
+    }
+
+    const frozenBalance = await this.calculateFrozenBalance(
+      offer.starkKeyA,
+      offer.aIsBuyingSynthetic ? AssetId.USDC : offer.syntheticAssetId
+    )
+
+    const offerValidated = validateInitialOffer(offer, positionA, frozenBalance)
+
+    if (!offerValidated) {
+      return { type: 'bad request', content: 'Your offer is invalid.' }
+    }
+
+    const createdAt = Timestamp(Date.now())
+
+    const id = await this.offerRepository.addInitialOffer({
+      createdAt,
+      ...offer,
+    })
+    return { type: 'created', content: { id } }
+  }
+
+  async acceptOffer(
+    initialOfferId: number,
+    acceptOfferData: Omit<ForcedTradeAcceptRecord, 'acceptedAt'>
+  ): Promise<ControllerResult> {
+    const positionB = await this.stateUpdateRepository.getPositionById(
+      acceptOfferData.positionIdB
+    )
+    const userRegistrationEventB =
+      await this.userRegistrationEventRepository.findByStarkKey(
+        acceptOfferData.starkKeyB
+      )
+
+    if (!positionB || !userRegistrationEventB) {
+      return { type: 'not found', content: 'Position does not exist.' }
+    }
+    const initialOffer = await this.offerRepository.findInitialOfferById(
+      initialOfferId
+    )
+
+    if (!initialOffer) {
+      return { type: 'not found', content: 'Offer does not exist.' }
+    }
+
+    const accceptOffer = await this.offerRepository.findAcceptOfferById(
+      initialOfferId
+    )
+
+    if (accceptOffer) {
+      return {
+        type: 'bad request',
+        content: 'Offer already accepted by a user.',
+      }
+    }
+
+    const frozenBalance = await this.calculateFrozenBalance(
+      acceptOfferData.starkKeyB,
+      initialOffer.aIsBuyingSynthetic
+        ? initialOffer.syntheticAssetId
+        : AssetId.USDC
+    )
+
+    const offerValidated = validateAcceptOffer(
+      initialOffer,
+      acceptOfferData,
+      positionB,
+      userRegistrationEventB.ethAddress,
+      frozenBalance
+    )
+
+    if (!offerValidated) {
+      return { type: 'bad request', content: 'Your offer is invalid.' }
+    }
+
+    const acceptedAt = Timestamp(Date.now())
+    await this.offerRepository.addAcceptOffer(initialOfferId, {
+      acceptedAt,
+      ...acceptOfferData,
+    })
+
+    return { type: 'success', content: 'Accept offer was submitted.' }
+  }
+
+  async calculateFrozenBalance(starkKey: StarkKey, assetId: AssetId) {
```

Comment:
I like the idea of this but this calculation actually has two problems:
1. I think that it makes sense for a user to create multiple offers for the same asset at different prices / amounts. Even if they only have 1000 units they should be able to create offers up to that amount.
2. When selling / buying user's USDC balance doesn't matter. Trading usually does not affect the total value of an account.

Reply:
So actually it's not necessary to check the `frozenBalance`